File: docs/stack/2296_hard/text_editor.cpp
#include <deque>
#include <string>
#include <vector>
using namespace std;

class TextEditor // LeetCode Q.2296.
{
private:
    deque<char> cursorLeftSide, cursorRightSide;

    string sliceCursorLeftTail() {
        string text = "";

        int startIdx = 0;
        if (cursorLeftSide.size() > 10)
            startIdx += cursorLeftSide.size() - 10;

        for (int idx = startIdx; idx < cursorLeftSide.size(); idx++)
            text += cursorLeftSide[idx];

        return text;
    }

public:
    TextEditor() {}

    void addText(string text) {
        for (const auto& letter : text)
            cursorLeftSide.push_back(letter);
    }

    int deleteText(int k) {
        int deletionsCount = cursorLeftSide.size();
        if (k < deletionsCount)
            deletionsCount = k;

        while (k > 0 && !cursorLeftSide.empty()) {
            cursorLeftSide.pop_back();
            k--;
        }

        return deletionsCount;
    }

    string cursorLeft(int k) {
        while (k > 0 && !cursorLeftSide.empty()) {
            char letter = cursorLeftSide.back();
            cursorRightSide.push_front(letter);
            cursorLeftSide.pop_back();
            k--;
        }

        return sliceCursorLeftTail();
    }

    string cursorRight(int k) {
        while (k > 0 && !cursorRightSide.empty()) {
            char letter = cursorRightSide.front();
            cursorLeftSide.push_back(letter);
            cursorRightSide.pop_front();
            k--;
        }

        return sliceCursorLeftTail();
    }
};

File: docs/stack/2296_hard/text_editor.py
from collections import deque


class TextEditor:
    def __init__(self) -> None:
        self.cursor_left_side: deque[str] = deque([])
        self.cursor_right_side: deque[str] = deque([])

    def _slice_cursor_left_tail(self) -> str:
        start_idx = max(0, len(self.cursor_left_side) - 10)

        text = ""
        for idx in range(start_idx, len(self.cursor_left_side)):
            text += self.cursor_left_side[idx]

        return text

    def add_text(self, text: str) -> None:
        self.cursor_left_side.extend(list(text))

    def delete_text(self, k: int) -> int:
        deletions_count = min(k, len(self.cursor_left_side))

        while k > 0 and self.cursor_left_side:
            self.cursor_left_side.pop()
            k -= 1

        return deletions_count

    def cursor_left(self, k: int) -> str:
        while k > 0 and self.cursor_left_side:
            char = self.cursor_left_side.pop()
            self.cursor_right_side.appendleft(char)
            k -= 1

        return self._slice_cursor_left_tail()

    def cursor_right(self, k: int) -> str:
        while k > 0 and self.cursor_right_side:
            char = self.cursor_right_side.popleft()
            self.cursor_left_side.append(char)
            k -= 1

        return self._slice_cursor_left_tail()
